guard/src/rules/eval_context.rs

```diff
@@ -1174,22 +1175,74 @@ impl<'value, 'loc: 'value> EvalContext<'value, 'loc> for RootScope<'value, 'loc>
     }
 }
 
-pub(crate) fn validate_number_of_params(name: &str, num_args: usize) -> Result<()> {
-    let expected_num_args = match name {
-        "join" => 2,
-        "substring" | "regex_replace" => 3,
-        "count" | "json_parse" | "to_upper" | "to_lower" | "url_decode" | "parse_string"
-        | "parse_boolean" | "parse_float" | "parse_int" | "parse_char" => 1,
-        _ => {
-            return Err(Error::ParseError(format!(
-                "no such function named {name} exists"
-            )));
+#[derive(Debug)]
+pub(crate) enum Function {
+    Join,
+    Substring,
+    RegexReplace,
+    Count,
+    JsonParse,
+    ToUpper,
+    ToLower,
+    UrlDecode,
+    ParseString,
+    ParseBoolean,
+    ParseFloat,
+    ParseInt,
+    ParseChar,
+}
+
+impl TryFrom<&str> for Function {
+    type Error = Error;
+
+    fn try_from(name: &str) -> std::result::Result<Self, Self::Error> {
+        match name {
+            "join" => Ok(Function::Join),
+            "substring" => Ok(Function::Substring),
+            "regex_replace" => Ok(Function::RegexReplace),
+            "count" => Ok(Function::Count),
+            "json_parse" => Ok(Function::JsonParse),
+            "to_upper" => Ok(Function::ToUpper),
+            "to_lower" => Ok(Function::ToLower),
+            "url_decode" => Ok(Function::UrlDecode),
+            "parse_string" => Ok(Function::ParseString),
+            "parse_boolean" => Ok(Function::ParseBoolean),
+            "parse_float" => Ok(Function::ParseFloat),
+            "parse_int" => Ok(Function::ParseInt),
+            "parse_char" => Ok(Function::ParseChar),
+            _ => Err(Error::ParseError(format!(
+                "No function with the name '{name}' exists.",
+            ))),
         }
+    }
+}
+
+pub(crate) fn validate_number_of_params(func: &Function, num_args: usize) -> Result<()> {
+    let valid_num_args = match func {
+        Function::Join => vec![2],
+        Function::Substring | Function::RegexReplace => vec![3],
+        Function::Count
+        | Function::JsonParse
+        | Function::ToUpper
+        | Function::ToLower
+        | Function::UrlDecode
+        | Function::ParseString
+        | Function::ParseBoolean
+        | Function::ParseFloat
+        | Function::ParseInt
+        | Function::ParseChar => vec![1],
     };
 
-    if expected_num_args != num_args {
+    if !valid_num_args.contains(&num_args) {
         return Err(Error::ParseError(format!(
-            "{name} function requires {expected_num_args} arguments be passed, but received {num_args}"
+            "{:?} function requires {} arguments be passed, but received {}",
```

Comment:
`{:?}` is the debug syntax, we wouldnt want to use that for the name, we should implement the display trait here. 

```
impl Display for Function {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let name = match self {
            Function::Join => "join",
            Function::Substring => "substring",
            Function::RegexReplace => "regex_replace",
            Function::Count => "count",
            Function::JsonParse => "json_parse",
            Function::ToUpper => "to_upper",
            Function::ToLower => "to_lower",
            Function::URLDecode => "url_decode",
            Function::ParseString => "parse_string",
            Function::ParseBoolean => "parse_boolean",
            Function::ParseFloat => "parse_float",
            Function::ParseInt => "parse_int",
            Function::ParseChar => "parse_char",
        };
        write!(f, "{}", name)
    }
}
```